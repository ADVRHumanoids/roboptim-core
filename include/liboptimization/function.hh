// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Function class.
 */

#ifndef OPTIMIZATION_FUNCTION_HH
# define OPTIMIZATION_FUNCTION_HH
# include <limits>
# include <utility>

# include <boost/numeric/ublas/matrix.hpp>
# include <boost/numeric/ublas/vector.hpp>

# include <liboptimization/fwd.hh>

namespace optimization
{
  namespace ublas = boost::numeric::ublas;

  /// \brief Define a mathematical function.
  class Function
  {
  public:
    /// Values type.
    typedef double value_type;
    /// Size type.
    typedef unsigned size_type;

    /// Vector type.
    typedef ublas::vector<value_type> vector_t;

    /// Matrix type.
    typedef ublas::matrix<value_type> matrix_t;

    /// Bound type (lower, upper).
    /// Use -infinity / +infinity to disable a bound.
    typedef std::pair<value_type, value_type> bound_t;
    /// Vector of bound.
    typedef std::vector<bound_t> bounds_t;

    /// Scale vector.
    typedef std::vector<value_type> scales_t;

    /// Function arity.
    const size_type n;
    /// Value that symbolizes infinity.
    const value_type infinity;

    /// Constructor.
    /// \param n function arity
    /// \param infinity value that encodes infinity
    Function (size_type n, value_type infinity =
              std::numeric_limits<value_type>::infinity ()) throw ();

    /// Destructor.
    virtual ~Function () throw ();

    /// Function.
    virtual value_type operator () (const vector_t&) const throw () = 0;

    /// Construct a bound from a lower and upper bound.
    bound_t makeBound (value_type, value_type) const throw ();
    /// Construct an infinite bound.
    bound_t makeBound () const throw ();
    /// Construct a bound from a lower bound.
    bound_t makeLowerBound (value_type) const throw ();
    /// Construct a bound from an upper bound.
    bound_t makeUpperBound (value_type) const throw ();

    /// Function bounds.
    bound_t bound;

    /// Arguments bounds.
    bounds_t argBounds;

    /// Result scale.
    double scale;
    /// Arguments' scales.
    scales_t argScales;
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_FUNCTION_HH
