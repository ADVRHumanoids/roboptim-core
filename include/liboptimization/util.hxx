// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures (templated) implementation.
 */
#ifndef OPTIMIZATION_UTIL_HXX
# define OPTIMIZATION_UTIL_HXX
# include <boost/variant/apply_visitor.hpp>

# include <liboptimization/function.hh>

namespace optimization
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }


    struct GenVariantPrintVisitor : public boost::static_visitor<void>
    {
      GenVariantPrintVisitor (std::ostream& o)
        : o_ (o)
      {}

      template <typename T>
      void operator () (const T& value)
      {
        o_ << value;
      }

      template <typename T>
      void operator () (const T* value)
      {
        if (!value)
          o_ << "Null pointer";
        else
          o_ << *value;
      }

    private:
      std::ostream& o_;
    };

  }; // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const boost::variant<T1, T2>& var)
  {
    detail::GenVariantPrintVisitor pv (o);
    boost::apply_visitor (pv, var);
    return o;
  }

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace optimization.

#endif //! OPTIMIZATION_UTIL_HXX
