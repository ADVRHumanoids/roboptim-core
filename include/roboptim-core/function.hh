// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Function class.
 */

#ifndef OPTIMIZATION_FUNCTION_HH
# define OPTIMIZATION_FUNCTION_HH
# include <iostream>
# include <limits>
# include <utility>

# include <boost/numeric/ublas/matrix.hpp>
# include <boost/numeric/ublas/vector.hpp>

# include <roboptim-core/fwd.hh>

namespace roboptim
{
  namespace ublas = boost::numeric::ublas;

  /// \brief Define a mathematical function.
  class Function
  {
  public:
    /// Values type.
    typedef double value_type;
    /// Size type.
    typedef unsigned size_type;

    /// Vector type.
    typedef ublas::vector<value_type> vector_t;

    /// Matrix type.
    typedef ublas::matrix<value_type> matrix_t;

    /// Problem dimension.
    const size_type n;

    /// Result dimension.
    const size_type m;

    /// Get the value that symbolizes infinity.
    static const value_type infinity () throw ()
    {
      return std::numeric_limits<Function::value_type>::infinity ();
    }

    /// Constructor.
    /// \param n function arity
    /// \param m result size
    Function (size_type n, size_type m = 1) throw ();

    /// Destructor.
    virtual ~Function () throw ();

    /// Function.
    virtual vector_t operator () (const vector_t&) const throw () = 0;

    /// Display function type.
    virtual std::ostream& print (std::ostream&) const throw ();
  };

  std::ostream& operator<< (std::ostream&, const Function&);
} // end of namespace roboptim

#endif //! OPTIMIZATION_FUNCTION_HH
