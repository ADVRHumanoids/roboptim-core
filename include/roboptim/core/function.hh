// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Function class.
 */

#ifndef ROBOPTIM_CORE_FUNCTION_HH
# define ROBOPTIM_CORE_FUNCTION_HH
# include <iostream>
# include <limits>
# include <utility>

# include <boost/numeric/ublas/matrix.hpp>
# include <boost/numeric/ublas/vector.hpp>
# include <boost/tuple/tuple.hpp>

# include <roboptim/core/fwd.hh>

namespace roboptim
{
  namespace ublas = boost::numeric::ublas;

  /// \addtogroup roboptim_function
  /// @{

  /// \brief Define an abstract mathematical function (\f$C^0\f$).
  ///
  /// A function is an object that can be evaluated for a given
  /// point.
  ///
  /// \f[ f : x \rightarrow f(x) \f]
  /// \f$x \in \mathbb{R}^n\f$, \f$f(x) \in \mathbb{R}^m\f$ where
  /// \f$n\f$ is the input size and \f$m\f$ is the output size.
  ///
  /// Functions are pure immutable objects: evaluating a function
  /// twice at a given point <b>must</b> give the same result.
  class Function
  {
  public:
    /// \brief Values type.
    ///
    /// Represents the numerical type (i.e. float, double, int...)
    /// used for computations.
    typedef double value_type;

    /// \brief Size type.
    ///
    /// This type is used to represent sizes, indexes, etc.
    typedef std::size_t size_type;

    /// \brief Basic vector type.
    ///
    /// This basic vector type is used each time a vector of values
    /// is required.
    ///
    /// \attention It is good practice in RobOptim to rely on this type
    /// when a vector of values is needed instead of relying on a particular
    /// implementation.
    typedef ublas::vector<value_type> vector_t;

    /// \brief Basic matrix type.
    ///
    /// This basic matrix type is used each time a two dimensional
    /// matrix of values is needed.
    ///
    /// \attention It is good practice in RobOptim to rely on this type
    /// when a matrix of values is needed instead of relying on a particular
    /// implementation.
    typedef ublas::matrix<value_type> matrix_t;

    /// \brief Type of a function evaluation result.
    typedef vector_t result_t;

    /// \brief Type of a function evaluation argument.
    typedef vector_t argument_t;


    /// \brief Get the value that symbolizes positive infinity.
    /// \return representation of positive infinity in the value type
    static const value_type infinity () throw ()
    {
      return std::numeric_limits<Function::value_type>::infinity ();
    }

    /// Interval
    /// \{

    /// \brief Interval type (lower, upper).
    /// Use negative or positive infinity to respectively disable the
    /// lower or upper bound.
    typedef std::pair<value_type, value_type> interval_t;

    /// \brief Vector of intervals.
    typedef std::vector<interval_t> intervals_t;

    /// \brief Construct an interval from a lower and upper bound.
    /// \param l lower bound
    /// \param u upper bound
    /// \return interval representing \f$[l, u]\f$
    static interval_t makeInterval (value_type l, value_type u) throw ()
    {
      assert (l <= u);
      return std::make_pair (l, u);
    }

    /// \brief Construct an infinite interval.
    /// \return interval representing \f$[-\infty, +\infty]\f$
    static interval_t makeInfiniteInterval () throw ()
    {
      return std::make_pair (-Function::infinity (), Function::infinity  ());
    }

    /// \brief Construct an interval from a lower bound.
    /// \param l lower bound
    /// \return interval representing \f$[l, +\infty]\f$
    static interval_t makeLowerInterval (value_type u) throw ()
    {
      return makeInterval (-Function::infinity  (), u);
    }

    /// \brief Construct an interval from an upper bound.
    /// \param u upper bound
    /// \return interval representing \f$[-\infty, u]\f$
    static interval_t makeUpperInterval (value_type l) throw ()
    {
      return makeInterval (l, Function::infinity  ());
    }

    /// \brief Get the lower bound of an interval
    /// \param interval accessed interval
    /// \return lower bound of the interval
    static double getLowerBound (const interval_t& interval) throw ()
    {
      return interval.first;
    }

    /// \brief Get the upper bound of an interval
    /// \param interval accessed interval
    /// \return upper bound of the interval
    static double getUpperBound (const interval_t& interval) throw ()
    {
      return interval.second;
    }

    /// \}

    /// Discrete interval
    /// \{

    /// \brief Types representing a discrete interval.
    /// A discrete interval is a triplet of values:
    /// - lower bound,
    /// - upper bound,
    /// - step.
    typedef boost::tuple<value_type,
			 value_type,
			 value_type> discreteInterval_t;


    /// \brief Construct a discrete interval.
    ///
    /// \param min miminum value of the interval
    /// \param max maxinum value of the interval
    /// \param step discretization step
    static discreteInterval_t makeDiscreteInterval (value_type min,
						    value_type max,
						    value_type step)
    {
      return discreteInterval_t (min, max, step);
    }

    /// \brief Get the lower bound of a discrete interval
    ///
    /// \param interval accessed discrete interval
    /// \return lower bound of the discrete interval
    static double getLowerBound (const discreteInterval_t& interval) throw ()
    {
      return boost::get<0> (interval);
    }

    /// \brief Get the upper bound of a discrete interval
    ///
    /// \param interval accessed discrete interval
    /// \return upper bound of the discrete interval
    static double getUpperBound (const discreteInterval_t& interval) throw ()
    {
      return boost::get<1> (interval);
    }

    /// \brief Get the upper step of a discrete interval
    ///
    /// \param interval accessed discrete interval
    /// \return upper step of the discrete interval
    static double getStep (const discreteInterval_t& interval) throw ()
    {
      return boost::get<2> (interval);
    }

    /// \}

    /// \brief Check the given result size is valid.
    ///
    /// \param result result that will be checked
    /// \return true if valid, false if not
    bool isValidResult (const result_t& result) const throw ()
    {
      return result.size () == outputSize ();
    }

    /// \brief Return the input size (i.e. argument's vector size).
    ///
    /// \return input size
    size_type inputSize () const throw ()
    {
      return inputSize_;
    }

    /// \brief Return the output size (i.e. result's vector size).
    ///
    /// \return input size
    size_type  outputSize () const throw ()
    {
      return outputSize_;
    }

    /// \brief Trivial destructor.
    virtual ~Function () throw ();

    /// \brief Evaluate the function at a specified point.
    ///
    /// The program will abort if the argument does not have the
    /// expected size.
    /// \param argument point at which the function will be evaluated
    /// \return computed result
    result_t operator () (const argument_t& argument) const throw ()
    {
      result_t result (outputSize ());
      result.clear ();
      (*this) (result, argument);
      return result;
    }

    /// \brief Evaluate the function at a specified point.
    ///
    /// The program will abort if the argument does not have the
    /// expected size.
    /// \param result result will be stored in this vector
    /// \param argument point at which the function will be evaluated
    void operator () (result_t& result, const argument_t& argument)
      const throw ()
    {
      assert (argument.size () == inputSize ());
      assert (isValidResult (result));
      this->impl_compute (result, argument);
      assert (isValidResult (result));
    }

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    /// \brief Concrete class constructor should call this constructor.
    ///
    /// \param inputSize function arity
    /// \param outputSize result size
    Function (size_type inputSize, size_type outputSize = 1) throw ();


    /// \brief Function evaluation.
    ///
    /// Evaluate the function, has to be implemented in concrete classes.
    /// \warning Do not call this function directly, call #operator() instead.
    /// \param result result will be stored in this vector
    /// \param argument point at which the function will be evaluated
    virtual void impl_compute (result_t&, const argument_t&)
      const throw () = 0;

  private:
    /// Problem dimension.
    const size_type inputSize_;

    /// Result dimension.
    const size_type outputSize_;
  };
  /// @}


  /// \brief Override operator<< to handle function display.
  ///
  /// \param o output stream used for display
  /// \param f function to be displayed
  /// \return output stream
  std::ostream& operator<< (std::ostream& o, const Function& f);


} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_FUNCTION_HH
