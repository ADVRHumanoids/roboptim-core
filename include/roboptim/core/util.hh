// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures declaration.
 */
#ifndef ROBOPTIM_CORE_UTIL_HH
# define ROBOPTIM_CORE_UTIL_HH
# include <boost/variant.hpp>
# include <roboptim/core/derivable-function.hh>

namespace roboptim
{
  namespace detail
  {
    /// \internal
    /// Copy the content of a uBLAS vector into a C array.
    void vector_to_array (Function::value_type* dst, const Function::vector_t& src);

    /// \internal
    /// Copy the content of a C array into a uBLAS vector.
    void array_to_vector (Function::vector_t& dst, const Function::value_type* src);

    /// \internal
    /// Merge gradients from several functions (each gradient is a line).
    /// The first line of the jacobian is the only one used.
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x);
  } // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const boost::variant<T1, T2>&);

  template <typename T>
  std::ostream& operator<< (std::ostream&, const std::vector<T>&);

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const std::pair<T1, T2>&);
} // end of namespace roboptim.

# include <roboptim/core/util.hxx>
#endif //! ROBOPTIM_CORE_UTIL_HH
