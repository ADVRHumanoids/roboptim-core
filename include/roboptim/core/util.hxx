// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_UTIL_HXX
# define ROBOPTIM_CORE_UTIL_HXX
# include <roboptim/core/function.hh>

namespace roboptim
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }
  } // end of namespace detail.

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_UTIL_HXX
