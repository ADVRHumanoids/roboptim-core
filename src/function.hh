// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \file src/function.hh
 *
 * \brief Declaration of the Function class.
 */

#ifndef OPTIMIZATION_FUNCTION_HH
# define OPTIMIZATION_FUNCTION_HH
# include <limits>
# include <utility>

# include <boost/numeric/ublas/matrix.hpp>
# include <boost/numeric/ublas/vector.hpp>
# include <boost/optional/optional.hpp>

# include <optimization-fwd.hh>

namespace optimization
{
  namespace ublas = boost::numeric::ublas;

  /// Define a mathematical function
  /// (and possibly a jacobian/hessian).
  struct Function
  {
    /// Values type.
    typedef double value_type;
    /// Size type.
    typedef unsigned size_type;

    /// Vector type.
    typedef ublas::vector<value_type> vector_t;

    /// Matrix type.
    typedef ublas::matrix<value_type> matrix_t;

    /// Gradient type.
    typedef boost::optional<vector_t> gradient_t;
    /// Hessian type.
    typedef boost::optional<matrix_t> hessian_t;

    /// Bound type (lower, upper).
    /// Use -infinity / +infinity to disable a bound.
    typedef std::pair<value_type, value_type> bound_t;
    /// Vector of bound.
    typedef std::vector<bound_t> bounds_t;

    /// Function arity.
    const size_type n;
    /// Value that symbolizes infinity.
    const value_type infinity;

    /// Linearity tag.
    enum Linearity
      {
        LINEAR,
        QUADRATIC,
        NON_LINEAR
      };

    /// Constructor.
    /// \param n function arity
    /// \param infinity value that encodes infinity
    Function (size_type n, value_type infinity =
              std::numeric_limits<value_type>::infinity ()) throw ();

    /// Destructor.
    virtual ~Function () throw ();

    /// Function.
    virtual value_type operator () (const vector_t&) const throw () = 0;
    /// Gradient.
    virtual gradient_t gradient (const vector_t&) const throw ();
    /// Hessian.
    virtual hessian_t hessian (const vector_t&) const throw ();

    /// Function bounds.
    bound_t bound;

    /// Arguments bounds.
    bounds_t argBounds;

    /// Function linearity.
    Linearity linearity;
  };
} // end of namespace optimization

#endif //! OPTIMIZATION_FUNCTION_HH
