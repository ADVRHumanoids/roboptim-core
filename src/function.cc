// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Implementation of the Function class.
 */

#include <algorithm>
#include "liboptimization/function.hh"
#include "liboptimization/util.hh"

namespace optimization
{
  Function::Function (unsigned n, value_type inf) throw ()
    : n (n),
      infinity (inf),
      bound (std::make_pair (-inf, inf)),
      argBounds (n),
      scale (1.),
      argScales (n)
  {
    // A positive infinite is required.
    assert (inf > 0.);

    // Positive size is required.
    assert (n >= 0);

    // Initialize bound.
    for (bounds_t::iterator it = argBounds.begin (); it != argBounds.end (); ++it)
      *it = std::make_pair (-inf, inf);

    // Initialize scale.
    std::fill (argScales.begin (), argScales.end (), 1.);
  }

  Function::~Function () throw ()
  {
  }

  Function::bound_t
  Function::makeBound (value_type l, value_type u) const
    throw ()
  {
    return std::make_pair (l, u);
  }

  Function::bound_t
  Function::makeBound () const
    throw ()
  {
    return std::make_pair (-infinity, infinity);
  }

  Function::bound_t
  Function::makeUpperBound (value_type u) const
  throw ()
  {
    return makeBound (-infinity, u);
  }

  Function::bound_t
  Function::makeLowerBound (value_type l) const
  throw ()
  {
    return makeBound (l, infinity);
  }

  std::ostream&
  Function::print (std::ostream& o) const throw ()
  {
    return o << "Function" << std::endl
             << "Result bound: " << bound << std::endl
             << "Arguments bounds " << argBounds << std::endl
             << "Result scale: " << scale << std::endl
             << "Arguments scale: " << argScales;
  }

  std::ostream&
  operator<< (std::ostream& o, const Function& f)
  {
    return f.print (o);
  }
} // end of namespace optimization
