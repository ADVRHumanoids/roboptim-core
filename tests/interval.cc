// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>

#include <boost/tuple/tuple_io.hpp>

#include "common.hh"
#include <roboptim/core/function.hh>
#include <roboptim/core/util.hh>

using namespace roboptim;

struct MyFunctor
{
  void operator () (double t)
  {
    std::cout << t << std::endl;
  }
};

int run_test ()
{
  typedef Function::interval_t interval_t;
  typedef Function::discreteInterval_t discreteInterval_t;

  {
    interval_t i1 = Function::makeInterval (0., 5.);
    assert (Function::getLowerBound (i1) == 0.);
    assert (Function::getUpperBound (i1) == 5.);

    interval_t i2 = Function::makeInfiniteInterval ();
    double inf = Function::infinity ();
    // Just make sure inf != NaN.
    assert (inf == inf);
    assert (Function::getLowerBound (i2) == -inf);
    assert (Function::getUpperBound (i2) == inf);

    interval_t i3 = Function::makeLowerInterval (42.);
    assert (Function::getLowerBound (i3) == 42.);
    assert (Function::getUpperBound (i3) == inf);

    interval_t i4 = Function::makeUpperInterval (42.);
    assert (Function::getLowerBound (i4) == -inf);
    assert (Function::getUpperBound (i4) == 42.);

    std::cout << i1 << std::endl
	      << i2 << std::endl
	      << i3 << std::endl
	      << i4 << std::endl;
  }
  {
    discreteInterval_t i1 = Function::makeDiscreteInterval (0., 5., 1.);
    assert (Function::getLowerBound (i1) == 0.);
    assert (Function::getUpperBound (i1) == 5.);
    assert (Function::getStep (i1) == 1.);

    double inf = Function::infinity ();
    discreteInterval_t i2 = Function::makeDiscreteInterval (-inf, inf, .1);
    assert (Function::getLowerBound (i2) == -inf);
    assert (Function::getUpperBound (i2) == inf);
    assert (Function::getStep (i2) == .1);

    discreteInterval_t i3 = Function::makeDiscreteInterval
      (Function::makeInterval (1., 4.), .1);

    std::cout << i1 << std::endl
	      << i2 << std::endl
	      << i3 << std::endl;

    Function::foreach (i1, MyFunctor ());
  }

  return 0;
}

GENERATE_TEST ()
