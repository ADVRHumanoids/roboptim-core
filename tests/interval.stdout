(0, 5)
(-1.#INF, 1.#INF)
(42, 1.#INF)
(-1.#INF, 42)
(0 5 1)
(-1.#INF 1.#INF 0.1)
(1 4 0.1)
0
1
2
3
4
5
